llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -9996,7 +10000,13 @@ AArch64TargetLowering::LowerELFGlobalTLSAddress(SDValue Op,
 
   const GlobalAddressSDNode *GA = cast<GlobalAddressSDNode>(Op);
 
-  TLSModel::Model Model = getTargetMachine().getTLSModel(GA->getGlobal());
+  TLSModel::Model Model;
+  if (DAG.getMachineFunction()
+          .getInfo<AArch64FunctionInfo>()
+          ->hasELFSignedGOT())
+    Model = TLSModel::GeneralDynamic;
+  else
+    Model = getTargetMachine().getTLSModel(GA->getGlobal());
```

Comment:
It seems that moving `AArch64FunctionInfo *MFI = ...` declaration here from the `if (Model == TLSModel::LocalDynamic) { ... }` block below would improve readability.

Reply:
Thanks for suggestion, applied this approach

Reply:
Initially, I assumed this would make it possible to use ternary operator as initializer, though it is completely up to you.

Reply:
Agree, ternary looks better here, changed in dcdbcdd133b329fa33f64ce5a7f57b8fcf785e4b